### ISSUE TITLE
BUG: object dtype quicksort can segfault

### ISSUE THREAD

**Issue by jaimefrio** [numpy/numpy#5961]
Running the example from #3879 with current master segfaults:

```
>>> import numpy as np
>>> class Raiser(object):
...     def raises_anything(*args, **kwargs):
...         raise TypeError("SOMETHING ERRORED")
...     __eq__ = __ne__ = __lt__ = __gt__ = __ge__ = __le__ = raises_anything
...
>>> arr = np.array([Raiser() for n in range(10)] + list(range(10)))
>>> arr.sort()
/Users/jaimefrio/miniconda/envs/numpydev/bin/python.app: line 3: 41998 Segmentation fault: 11  /Users/jaimefrio/miniconda/envs/numpydev/python.app/Contents/MacOS/python "$@"
```

This happens because of the changes introduced in #5670. We now use the same quicksort algorithm for objects as for other numeric types, instead of the stdlib's `qsort`. Our algorithm sorts the first, last and middle items of the array, both to choose the middle one as the pivot, and to have the first and last items as guards against run away pointers, which lets us avoid extra checks in the innermost loop.

Because on failure `PyObject_Compare` returns -1, and we do no checking if an error was raised in `OBJECT_compare`, these end guards do not serve their purpose anymore, hence the runaway pointer and the segfault.

I see two potential solutions:

 1. Modify `npy_quicksort`'s inner loop to do explicit inbounds checks before dereferencing a pointer.
 2. Modify `OBJECT_compare` to check if `PyErr_Occurred`, and return 0 rather than -1. This would unify the behavior with Python 3, where `PyObject_RichCompareBool` is called with `Py_LT`, then with `Py_GT`, and if none succeeds it returns 0 without explicitly checking with `Py_EQ`.

I prefer option 2.

**Comment by charris** [comment]
Option 2. looks better, but are we still able to signal that an error occurred? We should try to bail as soon as we see an error. Note that if we are calling python object methods, the tightness of the inner loop may not matter that much.

**Comment by jaimefrio** [comment]
Well. there isn't any early bail-out in place right now, which is part of the problem.

We can't simply start checking if `PyErr_Occurred` in the code of the generic `npy_quicksort`, because it would also be used by dtypes with no objects that today would release the GIL when sorting. If we want to bail out early (which seems a very reasonable thing to me), we either need to not release the GIL when calling this function, regardless of whether the type needs it or not (which may penalize a hypothetical user dtype), or create a `npy_quicksort_nogil` and use one or the other depending on `NPY_NEEDS_API` being set or not.

Also, git blames you for adding the check starting [here](https://github.com/numpy/numpy/blob/master/numpy/core/src/multiarray/arraytypes.c.src#L2607) to `OBJECT_compare` as part of 6951bcd, and I can't figure out the purpose of all that, do you remember?

**Comment by charris** [comment]
Hmm, I think the generic sorts should always hold the GIL, just in case. That choice used to be made at the function level in numarray.

Like the politicians say, I don't remember the purpose, but is looks like it was to avoid failure and push the check off to PyErr_Occurred. I was thinking another way to handle that in the comparison would be an error return, something like
```
int noerror = 1

while (LT(obj1, obj2, &noerror) && noerror) {whatever}
```

**Comment by charris** [comment]
Hmm, seems the example is fixed in 1.11.x
```
charris@fc [~]$ python segfault.py
segfault.py:8: RuntimeWarning: tp_compare didn't return -1 or -2 for exception
  arr.sort()
Exception TypeError: TypeError('SOMETHING ERRORED',) in <module 'threading' from '/usr/lib64/python2.7/threading.pyc'> ignored
```

**Comment by charris** [comment]
Fixed by #6269.
